File: core-mend/src/main/java/co/samco/mend4/core/crypto/CryptoProvider.java
package co.samco.mend4.core.crypto;

import co.samco.mend4.core.bean.EncodedKeyInfo;
import co.samco.mend4.core.bean.LogDataBlocksAndText;
import co.samco.mend4.core.exception.MalformedLogFileException;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.*;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.KeyPair;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.InvalidKeySpecException;

public interface CryptoProvider {
    int LENGTH_CODE_SIZE = 4;

    void encryptEncStream(RSAPublicKey publicKey, InputStream inputStream, OutputStream outputStream, String fileExtension)
            throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidAlgorithmParameterException,
            InvalidKeyException, InvalidKeySpecException, BadPaddingException, IllegalBlockSizeException, IOException;

    String decryptEncStream(RSAPrivateKey privateKey, InputStream inputStream, OutputStream outputStream)
            throws MalformedLogFileException, IOException, NoSuchPaddingException,
            NoSuchAlgorithmException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException,
            InvalidAlgorithmParameterException;

    void encryptLogStream(RSAPublicKey publicRsaKey, String text, OutputStream fos) throws NoSuchAlgorithmException,
            NoSuchPaddingException, InvalidAlgorithmParameterException, InvalidKeyException, IOException,
            BadPaddingException, IllegalBlockSizeException;

    void decryptLogStream(RSAPrivateKey privateKey, InputStream inputStream, PrintStream outputStream) throws IOException,
            InvalidAlgorithmParameterException, MalformedLogFileException,
            NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, NoSuchPaddingException, InvalidKeyException;

    boolean logHasNext(InputStream logInputStream, byte[] lc1Bytes) throws IOException, InvalidAlgorithmParameterException;

    LogDataBlocksAndText getNextLogTextWithDataBlocks(InputStream inputStream, RSAPrivateKey privateKey, byte[] lc1Bytes)
            throws IOException, MalformedLogFileException, IllegalBlockSizeException,
            InvalidKeyException, BadPaddingException, NoSuchAlgorithmException, NoSuchPaddingException,
            InvalidAlgorithmParameterException;

    RSAPublicKey getPublicKeyFromBytes(byte[] publicKey) throws NoSuchAlgorithmException, InvalidKeySpecException;

    RSAPrivateKey getPrivateKeyFromBytes(byte[] privateKey) throws NoSuchAlgorithmException, InvalidKeySpecException;

    KeyPair generateKeyPair() throws NoSuchAlgorithmException;

    KeyPair getKeyPairFromBytes(byte[] privateKey, byte[] publicKey) throws NoSuchAlgorithmException,
            InvalidKeySpecException;

    ///TODO we can and should do away with passcheck all together by generating and encrypting one on the fly
    boolean checkPassword(char[] password, String passCheck, String encryptedPassCheck)
            throws NoSuchAlgorithmException, InvalidKeySpecException, NoSuchPaddingException, InvalidAlgorithmParameterException,
            InvalidKeyException, BadPaddingException, IllegalBlockSizeException;

    byte[] decryptEncodedKey(char[] password, String encodedKey) throws InvalidKeySpecException,
            InvalidAlgorithmParameterException, NoSuchAlgorithmException, InvalidKeyException, NoSuchPaddingException,
            BadPaddingException, IllegalBlockSizeException;

    EncodedKeyInfo getEncodedKeyInfo(char[] password, String passCheck, KeyPair keyPair)
            throws NoSuchAlgorithmException, InvalidKeySpecException, NoSuchPaddingException,
            InvalidAlgorithmParameterException, InvalidKeyException, BadPaddingException, IllegalBlockSizeException;
}

File: core-mend/src/main/java/co/samco/mend4/core/crypto/CryptoProvider.kt
package co.samco.mend4.core.crypto

import co.samco.mend4.core.bean.LogDataBlocksAndText
import java.io.InputStream
import java.io.OutputStream
import java.io.PrintStream
import java.security.KeyPair
import java.security.PrivateKey
import java.security.PublicKey

sealed class UnlockResult {
    object Failure : UnlockResult()
    data class Success(val privateKey: PrivateKey) : UnlockResult()
}

interface CryptoProvider {
    /**
     * Encrypts the input stream and file extension and writes the encrypted data to the output stream.
     */
    suspend fun encryptEncStream(
        inputStream: InputStream,
        outputStream: OutputStream,
        fileExtension: String
    )

    /**
     * Decrypts the input stream and writes the decrypted data to the output stream.
     * @return the file extension of the decrypted file.
     */
    suspend fun decryptEncStream(
        privateKey: PrivateKey,
        inputStream: InputStream,
        outputStream: OutputStream
    ): String

    /**
     * Encrypts the text and writes the encrypted data to the output stream as a log entry.
     */
    suspend fun encryptLogStream(
        logText: String,
        outputStream: OutputStream
    )

    /**
     * Decrypts the input stream as a log and writes the decrypted data to the output stream.
     */
    suspend fun decryptLogStream(
        privateKey: PrivateKey,
        inputStream: InputStream,
        outputStream: PrintStream
    )

    /**
     * Reads 4 bytes from the input stream and puts them into [lc1Bytes].
     * @return true if the 4 bytes were successfully read, false otherwise.
     */
    fun logHasNext(logInputStream: InputStream, lc1Bytes: ByteArray): Boolean

    /**
     * Reads the next log entry from the input stream and decrypts it using the private key.
     * @return the decrypted log text with the full encrypted log bytes.
     */
    fun getNextLogTextWithDataBlocks(
        inputStream: InputStream,
        privateKey: PrivateKey,
        lc1Bytes: ByteArray
    ): LogDataBlocksAndText

    /**
     * Returns a [PrivateKey] from the given byte array. The byte array is expected to be PKCS8 encoded
     */
    fun getPrivateKeyFromBytes(privateKey: ByteArray): PrivateKey

    /**
     * Returns the public key
     */
    fun getPublicKey(): PublicKey

    /**
     * Creates a new [KeyPair].
     */
    fun generateKeyPair(): KeyPair

    /**
     * Returns a [KeyPair] from the given byte arrays. the [privateKey] should be PKCS8 encoded.
     * The [publicKey] should be X509 encoded.
     */
    fun getKeyPairFromBytes(privateKey: ByteArray, publicKey: ByteArray): KeyPair

    /**
     * Decrypts the stored private key using the given password and returns a PKCS8 encoded key
     * as a ByteArray or null if the password is incorrect.
     */
    fun unlock(password: CharArray): UnlockResult

    /**
     * Encrypts the given private key using the given password and store the key pair.
     */
    fun storeEncryptedKeys(password: CharArray, keyPair: KeyPair)
}